numpy/core/src/multiarray/descriptor.c

```diff
@@ -1266,30 +1263,34 @@ _convert_from_dict(PyObject *obj, int align)
                     "which is not divisible into the specified itemsize %d",
                     (int)new->alignment, (int)itemsize);
             Py_DECREF(new);
-            return NULL;
+            goto fail;
         }
         /* Set the itemsize */
         new->elsize = itemsize;
     }
 
     /* Add the metadata if provided */
-    metadata = Borrowed_PyMapping_GetItemString(obj, "metadata");
+    metadata = PyMapping_GetItemString(obj, "metadata");
 
     if (metadata == NULL) {
         PyErr_Clear();
     }
     else if (new->metadata == NULL) {
         new->metadata = metadata;
-        Py_XINCREF(new->metadata);
     }
     else if (PyDict_Merge(new->metadata, metadata, 0) == -1) {
         Py_DECREF(new);
-        return NULL;
+        Py_DECREF(metadata);
+        goto fail;
     }
     return new;
```

Comment:
I think metadata is leaked at the end of this if chain

Reply:
fixed